src/services/FirebaseService.ts

```diff
@@ -114,18 +135,50 @@ export class FirebaseService {
    * Requests custom token for the address from Firebase function and signs in to Firebase
    */
   private async signInToFirebase(address: string): Promise<UserCredential> {
+    if (firebaseAuth.currentUser && firebaseAuth.currentUser.uid === address) {
+      return;
+    }
+
     // Signs out from Firebase in case another user was authenticated
     // (could happen when user disconnect and connect a new wallet)
     await signOut(firebaseAuth);
 
-    // Requests a custom Firebase Token used for signing in to Firebase, from our Firebase Function
-    const token = await this.createCustomToken(address);
+    const messageToSign = await this.getMessageToSign(address);
 
-    // Firebase Authentication will be persisted in memory only
-    // that is on browser refresh the Firebase access token will be lost
-    await setPersistence(firebaseAuth, inMemoryPersistence);
+    const signature = await this.requestSignature(messageToSign);
+
+    if (!signature) {
+      throw new Error();
+    }
+
+    const token = await this.verifySignedMessageAndCreateCustomToken(address, signature);
+
+    // Firebase Authentication will be persisted in the browser storage (IndexedDB)
+    // user will be authenticated to Firebase as long as they don't clear the browser storage
+    // (or disconnect their wallet account, or switch to another account which will sign them out)
+    await setPersistence(firebaseAuth, browserLocalPersistence);
 
     // Signs in to Firebase with a given custom token
     return this.signInWithCustomToken(token);
   }
+
+  private async requestSignature(messageToSign: string): Promise<string> {
+    let signature: string;
+    try {
+      // Wait up to 30 seconds
+      signature = await Promise.race([
+        this.ethereumService.getDefaultSigner().signMessage(messageToSign),
+        Utils.timeout(30000),
+      ]);
+      this.eventAggregator.publish("handleSuccess", "Message was successfully signed");
+    } catch (error) {
+      const modal: IAlertModel = {
+        header: "Authentication failure",
+        message: "<p>You didn't sign the authentication message. You will only see public deals and you won't be able to edit your deals.</p>",
+      };
+      this.alertService.showAlert(modal);
```

Comment:
I don't think this service is the best place for UI interactions.  Is possible to do the user interactions in FirebaseDealsService,or even better, publish new events like "database.account.signed", passing: `{ account: Address, cancelled: boolean }`.  Then you can let app.ts subscribe and do the right things for the UI.